File: Report Generator/archive/report_generator.py
from collections import defaultdict
from pathlib import Path
from urllib.parse import quote
from typing import Dict, List, Tuple
from git_operations import get_line_creation_date


def generate_html_report(
    create_table_statements: Dict[str, Tuple[Path, int]],
    table_references: Dict[str, List[Tuple[Path, int, str]]],
    unused_tables: List[Tuple[str, Path]],
    table_columns: Dict[str, List[str]],
    column_references: Dict[str, Dict[str, List[Tuple[Path, int, str]]]],
    unused_columns: Dict[str, List[str]],
    queries: List[Dict],
    output_path: Path,
    repo_url: str,
    clone_path: Path,
    branch: str = "master",
) -> None:
    with output_path.open("w", encoding="utf-8") as f:
        f.write("<html><head><title>Database Table Usage Report</title></head><body>")
        f.write("<h1>Database Table Usage Report</h1>")

        # Summary Section
        f.write("<h2>Summary</h2>")
        f.write(f"<p>Total tables: {len(create_table_statements)}</p>")
        f.write(f"<p>Unused tables: {len(unused_tables)}</p>")
        if unused_tables:
            f.write("<ul>")
            for table, creation_file in unused_tables:
                creation_line = create_table_statements[table][1]
                relative_path = creation_file.relative_to(clone_path).as_posix()
                creation_date = get_line_creation_date(
                    clone_path, creation_file, creation_line
                )
                creation_date_text = (
                    f" (Created on: {creation_date})" if creation_date else ""
                )
                f.write(
                    f"<li><a href='{repo_url}/blob/{branch}/{quote(relative_path)}#L{creation_line}'>{table}</a> "
                    f"(Defined in: {relative_path}, Line: {creation_line}{creation_date_text})</li>"
                )
            f.write("</ul>")

        # Unused columns summary
        total_unused_columns = sum(len(cols) for cols in unused_columns.values())
        f.write(f"<p>Unused columns: {total_unused_columns}</p>")
        if total_unused_columns > 0:
            f.write("<ul>")
            for table, cols in unused_columns.items():
                creation_file, creation_line = create_table_statements[table]
                relative_path = creation_file.relative_to(clone_path).as_posix()
                f.write(
                    f"<li><strong>{table}</strong> (Defined in {relative_path}, line {creation_line}): "
                )
                f.write(", ".join(cols))
                f.write("</li>")
            f.write("</ul>")

        # Detailed Usage Section
        f.write("<h2>Detailed Table Usage</h2>")
        for table, references in table_references.items():
            f.write(f"<h3 id='{table}'>{table}</h3>")
            creation_data = create_table_statements.get(table)
            if creation_data:
                creation_file, creation_line = creation_data
                relative_path = creation_file.relative_to(clone_path).as_posix()
                creation_date = get_line_creation_date(
                    clone_path, creation_file, creation_line
                )
                creation_date_text = (
                    f" (Created on: {creation_date})" if creation_date else ""
                )
                f.write(
                    f"<p>Defined in: <a href='{repo_url}/blob/{branch}/{quote(relative_path)}#L{creation_line}'>"
                    f"{relative_path}, Line: {creation_line}</a>{creation_date_text}</p>"
                )
            else:
                f.write("<p>Definition file unknown.</p>")

            # Show columns
            f.write("<h4>Columns</h4>")
            f.write("<ul>")
            table_unused_cols = unused_columns.get(table, [])
            for col in table_columns.get(table, []):
                col_refs = column_references[table].get(col, [])
                if col_refs:
                    # Column is used
                    f.write(f"<li>{col} - used {len(col_refs)} times</li>")
                else:
                    # Unused column
                    f.write(f"<li>{col} - <strong>UNUSED</strong></li>")
            f.write("</ul>")

            if references:
                f.write("<h4>References</h4><ul>")
                for file_path, position, snippet in references:
                    relative_path = file_path.relative_to(clone_path).as_posix()
                    f.write(
                        f"<li>File: <a href='{repo_url}/blob/{branch}/{quote(relative_path)}#L{position}'>"
                        f"{relative_path}, Line: {position}</a>, Snippet: {snippet[:50]}</li>"
                    )
                f.write("</ul>")
            else:
                f.write("<p>No references found for this table.</p>")

            # Column references detailed section
            f.write("<h4>Column References</h4>")
            for col, col_refs in column_references[table].items():
                f.write(f"<h5>{col}</h5>")
                if col_refs:
                    f.write("<ul>")
                    for file_path, position, snippet in col_refs:
                        relative_path = file_path.relative_to(clone_path).as_posix()
                        f.write(
                            f"<li><a href='{repo_url}/blob/{branch}/{quote(relative_path)}#L{position}'>"
                            f"{relative_path}, Line {position}</a>: {snippet[:50]}</li>"
                        )
                    f.write("</ul>")
                else:
                    f.write("<p>No references found for this column.</p>")

        # Add Query Statistics Section
        generate_query_statistics_section(f, queries, clone_path, repo_url, branch)

        f.write("</body></html>")


def generate_query_statistics_section(
    f, queries: List[Dict], clone_path: Path, repo_url: str, branch: str
):
    """
    Generate an HTML section with detailed, commented statistics about the database queries:
    - Their type (SELECT, DELETE, INSERT, UPDATE, CREATE, etc.)
    - Their complexity (based on a simple heuristic)
    - Their distribution over the code base (which files contain the most queries)
    """

    # Compute queries by type
    queries_by_type = defaultdict(list)
    for q in queries:
        queries_by_type[q["type"]].append(q)

    # Compute complexity statistics
    # For each type, what is the average complexity?
    complexity_stats = {}
    for qtype, qlist in queries_by_type.items():
        complexities = [q["complexity"] for q in qlist]
        avg_complexity = sum(complexities) / len(complexities) if complexities else 0
        complexity_stats[qtype] = (len(qlist), avg_complexity)

    # Distribution by file: count queries per file
    queries_by_file = defaultdict(int)
    for q in queries:
        relative_path = q["file"].relative_to(clone_path).as_posix()
        queries_by_file[relative_path] += 1

    # Sort files by number of queries descending
    files_sorted = sorted(queries_by_file.items(), key=lambda x: x[1], reverse=True)

    f.write("<h2>Query Statistics</h2>")
    f.write("<p>Below are detailed statistics about the queries found in the code:</p>")

    # Queries by type
    f.write("<h3>Queries by Type</h3>")
    f.write("<ul>")
    for qtype, qlist in queries_by_type.items():
        count = len(qlist)
        f.write(f"<li>{qtype}: {count} queries</li>")
    f.write("</ul>")

    # Complexity statistics
    f.write("<h3>Complexity Statistics</h3>")
    f.write("<p>Complexity is a naive heuristic counting SQL keywords and length.</p>")
    f.write(
        "<table border='1'><tr><th>Query Type</th><th>Count</th><th>Average Complexity</th></tr>"
    )
    for qtype, (count, avg_complexity) in complexity_stats.items():
        f.write(
            f"<tr><td>{qtype}</td><td>{count}</td><td>{avg_complexity:.2f}</td></tr>"
        )
    f.write("</table>")

    # Distribution by file
    f.write("<h3>Distribution Over Code Base</h3>")
    if files_sorted:
        f.write("<p>Files with the most queries:</p>")
        f.write("<ol>")
        for filepath, qcount in files_sorted:
            f.write(f"<li>{filepath}: {qcount} queries</li>")
        f.write("</ol>")
    else:
        f.write("<p>No queries found.</p>")

    # Also, we can show a few example queries for each type
    f.write("<h3>Example Queries</h3>")
    for qtype, qlist in queries_by_type.items():
        f.write(f"<h4>{qtype}</h4>")
        # Show first 5 queries of this type
        examples = qlist[:5]
        if examples:
            f.write("<ul>")
            for ex in examples:
                relative_path = ex["file"].relative_to(clone_path).as_posix()
                f.write(
                    f"<li><a href='{repo_url}/blob/{branch}/{quote(relative_path)}#L{ex['line']}'>"
                    f"{relative_path}, line {ex['line']}</a>: {ex['snippet'][:100]}... "
                    f"(complexity: {ex['complexity']})</li>"
                )
            f.write("</ul>")
        else:
            f.write("<p>No examples.</p>")

File: Report Generator/src/main/java/com/example/analyzer/ReportGenerator.java
package com.example.analyzer;

import com.example.models.CodeReference;
import com.example.models.QueryInfo;
import com.example.models.TableDefinition;
import com.example.parsers.SQLParser;
import com.example.utils.HtmlUtils;
import java.io.BufferedWriter;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class ReportGenerator {
  // Store queries for complexity analysis
  public static List<QueryInfo> queries = new ArrayList<>();
  /**
   * Generates an HTML report summarizing the analysis.
   *
   * @param outputPath    The path to the output HTML file.
   * @param repoWebUrl    The web URL of the repository (for linking in the
   *     report).
   * @param repoDir       The directory of the cloned repository.
   * @param unusedTables  A list of unused tables.
   * @param unusedCols    A map of unused columns per table.
   * @param branch        The default branch of the repository.
   * @throws IOException If an I/O error occurs.
   */
  public static void generateHtmlReport(Path outputPath, String repoWebUrl,
                                        Path repoDir,
                                        List<TableDefinition> unusedTables,
                                        Map<String, List<String>> unusedCols,
                                        String branch) throws IOException {
    try (BufferedWriter w =
             Files.newBufferedWriter(outputPath, StandardCharsets.UTF_8)) {
      w.write("<html><head><title>Database Table Usage "
              + "Report</title></head><body>");
      w.write("<h1>Database Table Usage Report</h1>");

      // Summary
      w.write("<h2>Summary</h2>");
      w.write("<p>Total tables: " +
              DatabaseUsageAnalyzer.createTableStatements.size() + "</p>");
      w.write("<p>Unused tables: " + unusedTables.size() + "</p>");
      if (!unusedTables.isEmpty()) {
        w.write("<ul>");
        for (TableDefinition td : unusedTables) {
          String link = HtmlUtils.buildFileLink(repoWebUrl, repoDir, td.file,
                                                td.lineNumber, branch);
          String dateStr =
              (td.creationDate != null)
                  ? td.creationDate.format(DateTimeFormatter.ISO_LOCAL_DATE)
                  : "Unknown";
          w.write("<li>" + td.tableName + " (Defined in: <a href=\"" + link +
                  "\">" + repoDir.relativize(td.file) + ": line " +
                  td.lineNumber + "</a>"
                  + ", Created on: " + dateStr + ")</li>");
        }
        w.write("</ul>");
      }

      int totalUnusedColumns =
          unusedCols.values().stream().mapToInt(List::size).sum();
      w.write("<p>Unused columns: " + totalUnusedColumns + "</p>");
      if (totalUnusedColumns > 0) {
        w.write("<ul>");
        for (Map.Entry<String, List<String>> e : unusedCols.entrySet()) {
          w.write("<li><strong>" + e.getKey() +
                  "</strong>: " + String.join(", ", e.getValue()) + "</li>");
        }
        w.write("</ul>");
      }

      // Detailed Table Usage
      w.write("<h2>Detailed Table Usage</h2>");
      for (String table :
           DatabaseUsageAnalyzer.createTableStatements.keySet()) {
        w.write("<h3 id='" + table + "'>" + table + "</h3>");
        TableDefinition def =
            DatabaseUsageAnalyzer.createTableStatements.get(table);
        String defLink = HtmlUtils.buildFileLink(repoWebUrl, repoDir, def.file,
                                                 def.lineNumber, branch);
        String dateStr =
            (def.creationDate != null)
                ? def.creationDate.format(DateTimeFormatter.ISO_LOCAL_DATE)
                : "Unknown";
        w.write("<p>Defined in: <a href=\"" + defLink + "\">" +
                repoDir.relativize(def.file) + ": line " + def.lineNumber +
                "</a>"
                + ", Created on: " + dateStr + "</p>");

        // Columns
        w.write("<h4>Columns</h4><ul>");
        List<String> cols = DatabaseUsageAnalyzer.tableColumns.getOrDefault(
            table, Collections.emptyList());
        for (String col : cols) {
          boolean used =
              DatabaseUsageAnalyzer.columnReferences.containsKey(table) &&
              DatabaseUsageAnalyzer.columnReferences.get(table).containsKey(
                  col) &&
              !DatabaseUsageAnalyzer.columnReferences.get(table)
                   .get(col)
                   .isEmpty();
          if (used) {
            w.write("<li>" + col + "</li>");
          } else {
            w.write("<li>" + col + " - <strong>UNUSED</strong></li>");
          }
        }
        w.write("</ul>");

        // References
        List<CodeReference> refs =
            DatabaseUsageAnalyzer.tableReferences.get(table);
        if (refs != null && !refs.isEmpty()) {
          Map<String, Integer> typeCounts = new LinkedHashMap<>();
          for (CodeReference cr : refs) {
            String refType = SQLParser.classifyReferenceSnippet(cr.snippet);
            typeCounts.put(refType, typeCounts.getOrDefault(refType, 0) + 1);
          }

          // Summary table
          w.write("<h4>Reference Summary</h4>");
          w.write("<table border='1' cellpadding='4' cellspacing='0'>");
          w.write("<tr><th>Type</th><th>Count</th></tr>");
          for (Map.Entry<String, Integer> entry : typeCounts.entrySet()) {
            w.write("<tr><td>" + entry.getKey() + "</td><td>" +
                    entry.getValue() + "</td></tr>");
          }
          w.write("</table>");

          // References listing
          w.write("<h4>References</h4><ul>");
          for (CodeReference cr : refs) {
            String refLink = HtmlUtils.buildFileLink(
                repoWebUrl, repoDir, cr.file, cr.lineNumber, branch);
            w.write("<li><a href=\"" + refLink + "\">" +
                    repoDir.relativize(cr.file) + ": line " + cr.lineNumber +
                    "</a> - " + HtmlUtils.escapeHtml(cr.snippet) + "</li>");
          }
          w.write("</ul>");

        } else {
          w.write("<p>No references found.</p>");
        }
        w.write("<hr>");
      }

      // Global Query Statistics
      generateQueryStatisticsSection(w, repoWebUrl, repoDir, branch);

      w.write("</body></html>");
    }
  }
  /**
   * Generates the Query Statistics section of the HTML report.
   *
   * @param w          The BufferedWriter to write to.
   * @param repoWebUrl The web URL of the repository (for linking in the
   *     report).
   * @param repoDir    The directory of the cloned repository.
   * @param branch     The branch name.
   * @throws IOException If an I/O error occurs.
   */
  private static void
  generateQueryStatisticsSection(BufferedWriter w, String repoWebUrl,
                                 Path repoDir, String branch)
      throws IOException {
    w.write("<h2>Query Statistics</h2>");

    // Queries by type
    Map<String, List<QueryInfo>> queriesByType = new HashMap<>();
    for (QueryInfo qi : queries) {
      queriesByType.computeIfAbsent(qi.type, t -> new ArrayList<>()).add(qi);
    }

    // Overall summary
    int totalQueryCount = queries.size();
    w.write("<p>Total queries found: " + totalQueryCount + "</p>");

    // Summarize queries by type
    w.write("<h3>Queries by Type</h3><ul>");
    for (Map.Entry<String, List<QueryInfo>> e : queriesByType.entrySet()) {
      w.write("<li><strong>" + e.getKey() +
              "</strong>: " + e.getValue().size() + "</li>");
    }
    w.write("</ul>");

    // Summarize complexity
    w.write("<h3>Complexity Statistics</h3>");
    w.write("<table border='1'><tr><th>Query "
            + "Type</th><th>Count</th><th>Average Complexity</th></tr>");
    for (Map.Entry<String, List<QueryInfo>> e : queriesByType.entrySet()) {
      double avg = e.getValue()
                       .stream()
                       .mapToInt(q -> q.complexity)
                       .average()
                       .orElse(0.0);
      w.write("<tr><td>" + e.getKey() + "</td><td>" + e.getValue().size() +
              "</td><td>" + String.format("%.2f", avg) + "</td></tr>");
    }
    w.write("</table>");

    // *** Distribution by File ***
    w.write("<h3>Distribution by File</h3>");
    // Group queries by file, then count them
    Map<Path, Long> countByFile = queries.stream().collect(
        Collectors.groupingBy(q -> q.file, Collectors.counting()));

    // Sort descending by count
    List<Map.Entry<Path, Long>> sortedByCount =
        countByFile.entrySet()
            .stream()
            .sorted((a, b) -> Long.compare(b.getValue(), a.getValue()))
            .toList();

    w.write("<table border='1'>");
    w.write("<tr><th>File</th><th>Query Count</th></tr>");
    for (Map.Entry<Path, Long> entry : sortedByCount) {
      Path file = entry.getKey();
      long count = entry.getValue();
      String fileLink =
          HtmlUtils.buildFileLink(repoWebUrl, repoDir, file, 1, branch);

      w.write("<tr>");
      w.write("<td><a href=\"" + fileLink + "\">" + repoDir.relativize(file) +
              "</a></td>");
      w.write("<td>" + count + "</td>");
      w.write("</tr>");
    }
    w.write("</table>");

    // Detailed usage by type
    w.write("<h3>Detailed Query Usage</h3>");
    for (Map.Entry<String, List<QueryInfo>> entry : queriesByType.entrySet()) {
      String qtype = entry.getKey();
      w.write("<h4 id='querytype_" + qtype + "'>" + qtype + " Queries</h4>");
      List<QueryInfo> typedQueries = entry.getValue();

      if (typedQueries.isEmpty()) {
        w.write("<p>No queries found for this type.</p>");
        continue;
      }

      w.write("<ul>");
      for (QueryInfo qi : typedQueries) {
        String fileLink = HtmlUtils.buildFileLink(repoWebUrl, repoDir, qi.file,
                                                  qi.lineNumber, branch);
        w.write("<li>");
        w.write("<a href=\"" + fileLink + "\">" + repoDir.relativize(qi.file) +
                ": line " + qi.lineNumber + "</a>");
        w.write(" - <strong>Complexity:</strong> " + qi.complexity);
        w.write("<br/>" + HtmlUtils.escapeHtml(qi.snippet));
        w.write("</li>");
      }
      w.write("</ul>");
    }
  }
}
